File: src/main/java/proto/games/matchmaker/domain/matchmaker/position/PositionTeam.java
package proto.games.matchmaker.domain.matchmaker.position;

import lombok.Getter;
import proto.games.matchmaker.domain.matchmaker.rank.RankTeam;
import proto.games.matchmaker.domain.player.Player;
import proto.games.matchmaker.domain.player.Position;
import proto.games.matchmaker.domain.player.Rank;

import java.util.*;

@Getter
public class PositionTeam {
    private final RankTeam rankTeam;
    private final List<Player> players;
    private final Set<Position> filledPositions;

    public PositionTeam(RankTeam rankTeam) {
        this.rankTeam = rankTeam;
        this.players = new ArrayList<>();
        this.filledPositions = new HashSet<>();
    }

    public Player getHead() {
        return players.get(0);
    }

    public Rank getHeadRank() {
        return rankTeam.get(0);
    }

    public boolean include(Player player) {
        if (isIncompatible(player)) {
            return false;
        }

        if (hasNoPosition(player)) {
            return false;
        }

        return insert(player);
    }

    public boolean populate(Player player) {
        if (isIncompatible(player)) {
            return false;
        }

        return insert(player);
    }

    private boolean isIncompatible(Player player) {
        if (players.contains(player)) {
            return true;
        }

        if (!rankTeam.contains(player.getRank())) {
            return true;
        }

        if (wrongRank(player)) {
            return true;
        }

        return false;
    }

    private boolean insert(Player player) {
        filledPositions.add(player.getPrimaryPosition());
        filledPositions.add(player.getSecondaryPosition());

        var inserted = players.add(player);
        Collections.sort(players);

        return inserted;
    }

    private boolean wrongRank(Player player) {
        long totalRank = rankTeam.stream()
                .filter(r -> r == player.getRank())
                .count();

        long totalPlayersByRank = players.stream()
                .filter(p -> p.getRank() == player.getRank())
                .count();

        if (totalRank <= totalPlayersByRank) {
            return true;
        }

        return false;
    }

    private boolean hasNoPosition(Player player) {
        Position primary = player.getPrimaryPosition();
        Position secondary = player.getSecondaryPosition();

        if(!filledPositions.contains(primary)){
            return false;
        }

        if(!filledPositions.contains(secondary)){
            return false;
        }

        return true;
    }

}

File: src/main/kotlin/proto/games/matchmaker/domain/matchmaker/position/PositionTeam.kt
package proto.games.matchmaker.domain.matchmaker.position

import proto.games.matchmaker.domain.matchmaker.rank.RankTeam
import proto.games.matchmaker.domain.player.Player
import proto.games.matchmaker.domain.player.Position
import proto.games.matchmaker.domain.player.Rank


class PositionTeam(private val rankTeam: RankTeam) {
    val players: MutableList<Player> = ArrayList()
    val filledPositions: MutableSet<Position> = HashSet()

    val head: Player?
        get() = players.getOrNull(0)
    val headRank: Rank?
        get() = rankTeam.getOrNull(0)

    fun include(player: Player): Boolean {
        return when {
            hasNoPosition(player) -> false
            else -> populate(player)
        }
    }

    fun populate(player: Player): Boolean {
        return when {
            isIncompatible(player) -> false
            else -> insert(player)
        }
    }

    private fun isIncompatible(player: Player): Boolean {
        return when {
            players.contains(player) -> true
            rankTeam.contains(player.rank).not() -> true
            else -> wrongRank(player)
        }
    }

    private fun insert(player: Player): Boolean {
        filledPositions.add(player.primaryPosition)
        filledPositions.add(player.secondaryPosition)
        val inserted = players.add(player)
        players.sort()
        return inserted
    }

    private fun wrongRank(player: Player): Boolean {
        val totalRank = rankTeam
            .filter { r: Rank -> r === player.rank }
            .count()

        val totalPlayersByRank = players
            .filter { p -> p.rank === player.rank }
            .count()

        return totalRank <= totalPlayersByRank
    }

    private fun hasNoPosition(player: Player): Boolean {
        val primary: Position = player.primaryPosition
        val secondary: Position = player.secondaryPosition

        return when{
            filledPositions.contains(primary).not() -> false
            else -> filledPositions.contains(secondary).not()
        }
    }
}